### PR TITLE
PR2006 review

### PR DESCRIPTION
**Pull request by mikee47**
See review / commit notes

### CODE REVIEW

File: Sming/Components/OtaUpgrade/OtaUpgradeStream.h [hunk 1]
```diff
@@ -19,6 +19,25 @@
 #include <sodium/crypto_secretstream_xchacha20poly1305.h>
 #endif
 
+/**
+ * @brief Error code values
+ */
+#define OTA_ERRORCODE_MAP(XX)                                                                                          \
```
Comment (aemseemann) [inline Sming/Components/OtaUpgrade/OtaUpgradeStream.h:25]: @mikee47: I noticed that the x-macro approach prevents doxygen from picking up the enum values and their documentation. Do you have any idea how to get this working? I guess the Doxyfile could be tweaked such that at least the values are picked up, but I have no idea for the documentation part. An attempt like the following will certainly not work, as the preprocessor strips comments before doing macro expansion.
```cpp
#define XX(tag, text, desc) tag##Error , ///< desc
		OTA_ERRORCODE_MAP(XX)
```

Comment (mikee47) [reply Sming/Components/OtaUpgrade/OtaUpgradeStream.h:25]: Using this technique just isn't compatible with doxygen. As the format is fairly standard it could probably be pre-processed using a script (python?) either into doxygen comment format or .rst. Whether the effort is justified is another matter... 

Introducing a code generation tool into the build system is another approach. I probably wouldn't bother just for documentation (since we can just look at the source files) but there might be other uses for it. I've used XML for this sort of thing in the past, for example generating class code from a database schema.

Comment (aemseemann) [reply Sming/Components/OtaUpgrade/OtaUpgradeStream.h:25]: Thanks for the confirmation. I feared there is no way with doxygen. I had similar issues with x-macros and static analysis tool justification syntax in the past... 

Preprocessing the source would probably involve something like libclang (probably the only sane solution to deal with the complexity of modern C++, doxygen also has an [option](http://www.doxygen.nl/manual/config.html#cfg_clang_assisted_parsing) to do its parsing based on clang). However, the additional complexity is certainly not justified for proper enum documentation. 

On the other hand, I am not convinced that in this particular case, the maintenance benefit of the X-macro outweighs the documentation drawback (the argument to look at the sources somewhat counteracts the documentation effort). Any recent compiler warns about missing enum values in a switch/case statement, so keeping such pieces of code consistent with the enum definition is not that hard or risky. 
Bottom line: Would you accept the review if I revert back to the original solution without the X-macro?

Comment (mikee47) [reply Sming/Components/OtaUpgrade/OtaUpgradeStream.h:25]: Sure, if a better way of doing this presents itself then we can revisit.

Comment (mikee47) [reply Sming/Components/OtaUpgrade/OtaUpgradeStream.h:25]: ESP IDF has an interesting approach https://github.com/espressif/esp-idf/blob/master/components/esp_common/src/esp_err_to_name.c

Comment (aemseemann) [reply Sming/Components/OtaUpgrade/OtaUpgradeStream.h:25]: Seems like one of the many workarounds for missing reflection support in C++, in particular enum member <-> string conversion. (At least we finally got a TS in C++20.)

However, I don't think the ESP IDF approach has any benefit in our case, since 
- the description strings are supposedly different from the enum values
- The compiler cannot warn about missing enum values.

File: Sming/Components/OtaUpgrade/OtaUpgradeStream.cpp [hunk 2]
```diff
@@ -14,18 +14,19 @@
 #ifdef OTA_ENCRYPTED
 #include <sodium/utils.h>
 #endif
+#include <FlashString/Array.hpp>
 
-// symbol provided by appcode (generated source file)
+// Keys defined by `appcode/keys.cpp` which is compiled with application
 #ifdef OTA_SIGNED
-extern "C" uint8_t OTAUpgrade_SignatureVerificationKey_P[] PROGMEM;
+DECLARE_FSTR_ARRAY(OTAUpgrade_SignatureVerificationKey, uint8_t);
```
Comment (aemseemann) [inline Sming/Components/OtaUpgrade/OtaUpgradeStream.cpp:21]: After some debugging, it seems `DECLARE_FSTR_ARRAY` is not compatible with `IMPORT_FSTR_ARRAY`, because the former declares a reference
```
extern const FSTR::Array<ElementType> /* --> */ & /* <-- */ name;
```
and the latter a "by-value" symbol:
```
extern "C" const FSTR::Array<ElementType> name;
```
This should probably be added to the documentation somehow.
Or how do you feel about extending `IMPORT_FSTR_ARRAY` like this:
```cpp
#define IMPORT_FSTR_ARRAY(name, ElementType, file) \
	IMPORT_FSTR_DATA(name##_imported, file) \
	extern "C" const FSTR::Array<ElementType> name##_imported; \
	const FSTR::Array<ElementType>& name = name##_imported
```

Comment (mikee47) [reply Sming/Components/OtaUpgrade/OtaUpgradeStream.cpp:21]: I did cover this issue with [IMPORT_FSTR_DATA](https://sming.readthedocs.io/en/latest/_inc/Sming/Components/FlashString/utility.html#c.IMPORT_FSTR_DATA) but didn't properly document the related macros.

Your suggestion of defining an additional reference makes a great deal of sense, simplifying the documentation and making it easier to use. Thank you! I'll update the library.

Comment (mikee47) [reply Sming/Components/OtaUpgrade/OtaUpgradeStream.cpp:21]: OK, I've updated the FlashString library. It's a pity we still need the C symbol for linkage...
